java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -933,9 +940,21 @@ public void visitMethodReference(MethodReferenceTree methodReferenceTree) {
         .map(samMethod -> (JavaType) samMethod.returnType().type())
         .ifPresent(samReturnType -> {
           JavaSymbol methodSymbol = (JavaSymbol) methodRefTree.method().symbol();
-          if (!"<init>".equals(methodSymbol.name) && methodSymbol.isMethodSymbol()) {
+          if (methodSymbol.isMethodSymbol()) {
             JavaType capturedReturnType = resolve.resolveTypeSubstitution(samReturnType, methodRefType);
             JavaType refinedReturnType = ((MethodJavaType) methodRefTree.method().symbolType()).resultType();
+            if ("<init>".equals(methodSymbol.name)) {
+              JavaType sanitizedCaptured = capturedReturnType;
+              if (refinedReturnType.symbol().isTypeSymbol() && !((JavaSymbol.TypeJavaSymbol) refinedReturnType.symbol()).typeParameters().scopeSymbols().isEmpty()) {
+                refinedReturnType = parametrizedTypeCache.getParametrizedTypeType(refinedReturnType.symbol, new TypeSubstitution());
+              }
+              if (sanitizedCaptured.isTagged(JavaType.TYPEVAR)) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)